### ISSUE TITLE
Database already exists exception - Database Name is always Case Sensitive even when it shouldn't be

### ISSUE THREAD

**Issue by RachelAmbler** [erikbra/grate#167]
When running a deployment I'm getting the following exception raised:
`Unhandled exception: Microsoft.Data.SqlClient.SqlException (0x80131904): Database 'RetractedDatabaseName' already exists. Choose a different database name.`

This is the command line:

`-cs "Data Source="RetractedServerName.test.local, 1433";Initial Catalog=RetractedDatabaseName;Trust Server Certificate=True;Authentication=ActiveDirectoryIntegrated" -ct 6000 --verbosity Debug -f Grate`

I'm not expecting it to try and create the database!

MacOS BigSur 11.6.1

I suspect this may be due to [grate/Configuration/GrateConfiguration Line 52](https://github.com/erikbra/grate/blob/main/grate/Configuration/GrateConfiguration.cs#L52)

```csharp
public bool CreateDatabase { get; init; } = true;
```

And I'm not finding a corresponding command line option to turn that off.

**Comment by wokket** [comment]
Grate is designed to run against existing databases just fine, that flag indicates whether to create a database if it _doesn't_ exist yet, so there's something deeper going on here...  Do you have a (suitably redacted) log trace you can upload showing the exists checks etc going on prior to trying to create?

**Comment by RachelAmbler** [comment]
Yeah - I can get at that and send it to you later tonight or early tomorrow. Thanks!

**Comment by RachelAmbler** [comment]
> Grate is designed to run against existing databases just fine, that flag indicates whether to create a database if it _doesn't_ exist yet, so there's something deeper going on here... Do you have a (suitably redacted) log trace you can upload showing the exists checks etc going on prior to trying to create?

So, what's the best way to generate this log trace? Just with `--Verbose Debug` or `Trace` - or someway else?

**Comment by wokket** [comment]
`--verbosity` flag is perfect.

**Comment by RachelAmbler** [comment]
```
Unhandled exception: Microsoft.Data.SqlClient.SqlException (0x80131904): Database 'RetractedDatabaseName' already exists. Choose a different database name.
   at Microsoft.Data.SqlClient.SqlConnection.OnError(SqlException exception, Boolean breakConnection, Action`1 wrapCloseInAction)
   at Microsoft.Data.SqlClient.SqlInternalConnection.OnError(SqlException exception, Boolean breakConnection, Action`1 wrapCloseInAction)
   at Microsoft.Data.SqlClient.TdsParser.ThrowExceptionAndWarning(TdsParserStateObject stateObj, Boolean callerHasConnectionLock, Boolean asyncClose)
   at Microsoft.Data.SqlClient.TdsParser.TryRun(RunBehavior runBehavior, SqlCommand cmdHandler, SqlDataReader dataStream, BulkCopySimpleResultSet bulkCopyHandler, TdsParserStateObject stateObj, Boolean& dataReady)
   at Microsoft.Data.SqlClient.SqlCommand.InternalEndExecuteNonQuery(IAsyncResult asyncResult, Boolean isInternal, String endMethod)
   at Microsoft.Data.SqlClient.SqlCommand.EndExecuteNonQueryInternal(IAsyncResult asyncResult)
   at Microsoft.Data.SqlClient.SqlCommand.EndExecuteNonQueryAsync(IAsyncResult asyncResult)
   at System.Threading.Tasks.TaskFactory`1.FromAsyncCoreLogic(IAsyncResult iar, Func`2 endFunction, Action`1 endAction, Task`1 promise, Boolean requiresSynchronization)
--- End of stack trace from previous location ---
   at grate.Migration.AnsiSqlDatabase.ExecuteNonQuery(DbConnection conn, String sql)
   at grate.Migration.AnsiSqlDatabase.ExecuteNonQuery(DbConnection conn, String sql)
   at grate.Migration.AnsiSqlDatabase.CreateDatabase()
   at grate.Migration.DbMigrator.CreateDatabase()
   at grate.Migration.GrateMigrator.CreateDatabaseIfItDoesNotExist(IDbMigrator dbMigrator)
   at grate.Migration.GrateMigrator.Migrate()
   at grate.Commands.MigrateCommand.<>c__DisplayClass0_0.<<-ctor>b__0>d.MoveNext()
--- End of stack trace from previous location ---
   at System.CommandLine.Invocation.CommandHandler.GetExitCodeAsync(Object value, InvocationContext context)
   at System.CommandLine.Invocation.ModelBindingCommandHandler.InvokeAsync(InvocationContext context)
   at System.CommandLine.Invocation.InvocationPipeline.<>c__DisplayClass4_0.<<BuildInvocationChain>b__0>d.MoveNext()
--- End of stack trace from previous location ---
   at System.CommandLine.Builder.CommandLineBuilderExtensions.<>c__DisplayClass23_0.<<UseParseErrorReporting>b__0>d.MoveNext()
--- End of stack trace from previous location ---
   at System.CommandLine.Builder.CommandLineBuilderExtensions.<>c__DisplayClass16_0.<<UseHelp>b__0>d.MoveNext()
--- End of stack trace from previous location ---
   at System.CommandLine.Builder.CommandLineBuilderExtensions.<>c__DisplayClass25_0.<<UseTypoCorrections>b__0>d.MoveNext()
--- End of stack trace from previous location ---
   at System.CommandLine.Builder.CommandLineBuilderExtensions.<>c.<<UseSuggestDirective>b__24_0>d.MoveNext()
--- End of stack trace from previous location ---
   at System.CommandLine.Builder.CommandLineBuilderExtensions.<>c__DisplayClass22_0.<<UseParseDirective>b__0>d.MoveNext()
--- End of stack trace from previous location ---
   at System.CommandLine.Builder.CommandLineBuilderExtensions.<>c__DisplayClass11_0.<<UseDebugDirective>b__0>d.MoveNext()
--- End of stack trace from previous location ---
   at System.CommandLine.Builder.CommandLineBuilderExtensions.<>c.<<RegisterWithDotnetSuggest>b__10_0>d.MoveNext()
--- End of stack trace from previous location ---
   at System.CommandLine.Builder.CommandLineBuilderExtensions.<>c__DisplayClass14_0.<<UseExceptionHandler>b__0>d.MoveNext()
ClientConnectionId:a4bbbadd-5b83-497b-accf-ffafef41d0a3
Error Number:1801,State:3,Class:16```

**Comment by wokket** [comment]
Thanks for that...  there was no other debug info logged prior to the exception?

You can see the logic in `CreateDatabaseIfItDoesNotExist` [here](https://github.com/erikbra/grate/blob/c36b3cb5d850db5ca683582ce507f876c2289549/grate/Migration/GrateMigrator.cs#L191-L205), and it's very specifically only creating if it fails the `Exists()` check.  

My suspicion is that we might have a case sensitivity issue or something in the [Exists logic](https://github.com/erikbra/grate/blob/c36b3cb5d850db5ca683582ce507f876c2289549/grate/Migration/AnsiSqlDatabase.cs#L112-L138) ... are you running a case sensitive collation?  Could you try again ensuring you match the case exactly?

**Comment by RachelAmbler** [comment]
I think it might indeed be the case - although not on the SQL Server side. Before I run the grate command I call a `sqlcmd` script where I pass the exact same Database name in to get the `Db_Id()` and it returns a value.

However the case in the name I'm sending doesn't match the one on SQL Server. Must be something in `grate` that's balking at the case.

**Comment by RachelAmbler** [comment]
Gonna guess that the `databases` collection is Case Sensitive? If so then `return databases.Contains(DatabaseName);` will fall over on the case.

**Comment by RachelAmbler** [comment]
So I've been able to get past this exception for now - as I mentioned I was already doing a check beforehand and getting the `Db_Id()` value. I changed that to do a `IsNull(Db_Name(Db_Id('${Database}')), '.')` which now returns back the same case name of the database as is stored on SQL Server, thus skating past the problem.

That said, this is still most defiantly a bug. Just not one that my Pull request would fix.

**Comment by wokket** [comment]
So just confirming - passing the database name with the came case as the server works, but fails on a casing difference?  And you have a non-case-sensitive collation running on the server?

**Comment by RachelAmbler** [comment]
That is correct. 

e.g. If my Database Name is HiFiSounds then

```sql
Select Db_Id('hifiSOUNDS'), Db_Id('HiFiSounds')
```

will return back the exact same underlying `database_id` value. Passing `HiFiSounds` to grate will work fine, but passing `hifiSOUNDS` will cause the aforementioned exception.

**Comment by RachelAmbler** [comment]
Getting a list of databases and looking for a match outside of the RDBMS isn't possibly the best approach. Sure you could make the collection case-insensitive but then that might cause issues when you DO have a Case Sensitive collection.

Would a better way be to do what I'm basically doing outside of `grate` and making a Database call to confirm existence - that way you'd always be operating under the auspices of the underlying system and not just an assumption.

**Comment by erikbra** [comment]
Hi, @RachelAmbler , sorry to chime in so late, but, being late paid off, you already seem to have solved the issue :)
I do think you're right, though. It would make sense to use the database itself to decide whether a database exists or not, because then we would honour any particularities around casing (and any other possible issues) that the particular database might have.

That said, I also think we are missing the option to _not_ create a database, even if it does not exist. One might argue whether it is a deficiency or a feature, but I see the argument for just wanting to fail if the database does not exist.

So, in summary, we have two new feature requests:

1. Add an option similar to RoundhousE's [`--donotcreatedatabase`](https://github.com/chucknorris/roundhouse/wiki/ConfigurationOptions#switches) flag
1. Make the check to see if a database exists implemented in the database itself, e.g. using `Db_Id` in SQL Server

Do we agree? 😃

**Comment by RachelAmbler** [comment]
@erikbra I'm having a hard time finding any negatives to your proposal.

**Comment by wokket** [comment]
@erikbra FYI this has just bitten me too so I'm going to get a PR to you as soon as I get a chance to fix the issue.
